v0/destinations/pinterest_tag/utils.js

```diff
@@ -0,0 +1,295 @@
+ * @returns
+ * Further Processing the user fields following the instructions of Pinterest Conversion API
+ * Ref: https://s.pinimg.com/ct/docs/conversions_api/dist/v3.html
+ */
+const processUserPayload = userPayload => {
+  let formatValue = "";
+  Object.keys(userPayload).forEach(key => {
+    switch (key) {
+      case "em":
+        formatValue = userPayload[key].toString().toLowerCase();
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "ph":
+      case "zp":
+        // zip fields should only contain digits
+        formatValue = userPayload[key].toString().replace(/[^0-9]/g, "");
+        if (key === "ph") {
+          // phone numbers should not contain leading zeros
+          formatValue = formatValue.replace(/^0+/, "");
+        }
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "ct":
+      case "st":
+      case "country":
+      case "ge":
+      case "db":
+      case "ln":
+      case "fn":
+      case "hashed_maids":
+        userPayload[key] = [sha256(userPayload[key])];
+        break;
+      default:
+        userPayload[key] = String(userPayload[key]);
+    }
+  });
+  return userPayload;
+};
+
+/**
+ *
+ * @param {*} message
+ * @returns opt_out status
+ *
+ */
+
+const deduceOptOutStatus = message => {
+  const adTrackingEnabled = message.context?.device?.adTrackingEnabled;
+  const attStatus = message.context?.device?.attStatus;
+  let opt_out;
+
+  // for android
+  if (isDefinedAndNotNull(adTrackingEnabled)) {
+    if (adTrackingEnabled === true) {
+      opt_out = false;
+    } else if (adTrackingEnabled === false) {
+      opt_out = true;
+    }
+  }
+
+  // for ios
+  if (isDefinedAndNotNull(attStatus)) {
+    opt_out = attStatus === 3 ? true : false;
+  }
+
+  return opt_out;
+};
+
+/**
+ *
+ * @param {*} message
+ * @returns
+ * Maps the required common parameters accross event types. Checks for the correct
+ * action source types and deduces opt_out status
+ * Ref: https://s.pinimg.com/ct/docs/conversions_api/dist/v3.html
+ */
+const processCommonPayload = message => {
+  const commonPayload = constructPayload(message, COMMON_CONFIGS);
+  const presentActionSource = commonPayload.action_source;
+  if (
+    presentActionSource &&
+    !VALID_ACTION_SOURCES.includes(presentActionSource.toLowerCase())
+  ) {
+    throw new CustomError(
+      "Action source must be one of 'app_android', 'app_ios', 'web', 'offline' ",
+      400
+    );
+  }
+
+  commonPayload.opt_out = deduceOptOutStatus(message);
+
+  return commonPayload;
+};
+
+/**
+ *
+ * @param {*} message
+ * @param {*} Config
+ * @returns
+ * For the few ecommerce events the mapping is like following:
+ * const ecomEventMaps = [
+    {
+      src: ["order completed"],
+      dest: "Checkout",
+    },
+    {
+      src: ["product added"],
+      dest: "AddToCart",
+    },
+    {
+      src: ["products searched", "product list filtered"],
+      dest: "Search",
+    },
+  ];
```

Comment:
Here, if someone sets "order completed" to "custom", still it gets set to "checkout". Implemented by following the device mode implementation. But shouldn't we provide user the facility to choose whatever event they would want to send? In that case the order of conditions will be a bit altered. We have to check for the webapp map first, if not there, we will check for the default map. WDYT @utsabc